Short transformations guide (#155)
thehyve/delphyne: docs/transformations.rst

    :local:
    :backlinks: none

Introduction

> -------------

Reviewer: The dashes should be the same length as the title

The delphyne :class:`.Wrapper` offers different ways to execute transformation scripts.
Basically, there are three options:

>  - Write a Python function, returning the results as `SQLAlchemy ORM objects <https://docs.sqlalchemy.org/en/14/orm/tutorial.html>`_.
>  - Write a SQL query, directly inserting into the target table.
>  - Write a Python function, creating a query using the `SQLAlchemy expression language <https://docs.sqlalchemy.org/en/14/core/tutorial.html>`_.

Reviewer: ```suggestion
    - Write a Python function, returning the results as `SQLAlchemy ORM objects <https://docs.sqlalchemy.org/en/14/orm/tutorial.html>`_.
    - Write a SQL query, directly inserting into the target table.
    - Write a Python function, creating a query using the `SQLAlchemy expression language <https://docs.sqlalchemy.org/en/14/core/tutorial.html>`_.
```

Reviewer: This worked fine, but somehow when generating the docs with sphinx I still get a complaint on the indentation level. I made it 4 spaces and the warning is gone.

>  - Write a Python function, creating a query using the `SQLAlchemy expression language <https://docs.sqlalchemy.org/en/14/core/tutorial.html>`_.

Reviewer: when generating the docs with sphinx I get a complaint, add an empty line before the first bullet point and it will go away

Reviewer: Another note: should the links point one step higher, to the ORM/core reference instead of directly to the tutorial?

> For both options the :class:`.Wrapper` has built-in methods that coordinate the execution.

Reviewer: ```suggestion
For all of these options, the :class:`.Wrapper` has built-in methods that coordinate the execution.
```

Reviewer: (cannot use both with > 2)

These methods, detailed below, use SQLAlchemy to handle database operations and log the execution statistics.
This provides a standardised interface independent of whether you are using Python or SQL for the transformation logic.

It is good practice to create a separate transformation file (either Python or SQL) for each table to table
transformation.
This improves maintainability and allows for delphyne creating a detailed mapping log.
In addition, the advised convention for the transformation file naming is ``<source_table>_to_<target_table>.py/sql``.

SQLAlchemy ORM

> ---------------

Reviewer: The dashes should be the same length as the title

The SQLAlchemy ORM approach allows the use of 'simple' Python to write transformations.
This function takes the wrapper as input and produces OMOP records as SQLAlchemy ORM objects.

> It can return these objects in two different ways (see below).

Reviewer: I think you can remove this line, it's says again below that there's 2 ways of executing

The convention is to put the transformation scripts in the `src/main/python/transformation` folder.
`See delphyne-template for examples. <https://github.com/thehyve/delphyne-template/tree/master/src/main/python/transformation>`_

> There are two ways of execution the transformation, each requiring a different return type.

Reviewer: "Two ways of executiNG" ?

The first mode, used by :meth:`.Wrapper.execute_transformation`.

> This expects the transformation function to return the full list of ORM objects at once.

Reviewer: The first sentence ends strangely with a dot, I think you can merge them ("[...] Wrapper.execute_transformation expects the transformation function..")

All returned records are committed to the database at the same time.

.. code-block:: python

    def my_transformation(wrapper):
        # ...read source data...
        records_to_insert = []
        for row in source:
            # ...transformation logic...

>             omop_record = wrapper.cdm.Observation(

Reviewer: not sure about that "omop_record".. just "record" or "cdm_record"? :P

                person_id=...,
                observation_concept_id=...,
                ...
            )
            records_to_insert.append(records_to_insert)
        return records_to_insert

While this is straightforward, it does require to store all records in memory.
For large tables there is a 'batch' mode :meth:`.Wrapper.execute_batch_transformation`.
This expects the transformation function to yield the ORM objects one at a time.

> The wrapper will commits the yielded records to the database in batches of given size.

Reviewer: Important: add (in parenthesis?) that this is decided via the `batch_size` option

.. code-block:: python

    def my_batch_transformation(wrapper):
        # ...read source data...
        for row in source:
            # ...transformation logic...
            omop_record = wrapper.cdm.Observation(
                person_id=...,
                observation_concept_id=...,
                ...
            )
            yield omop_record

> In the wrapper these two transformations can be called like this:

Reviewer: Suggestion: in other places we specifically say to add calls to the Wrapper's **run method**, so you could change it to make it very explicit

Author: @alepev Wait a minute, but the delphyne-template has these execute calls in the `transform()` method... 👀 
https://github.com/thehyve/delphyne-template/blob/master/src/main/python/wrapper.py

Reviewer: True, and in fact in a separate PR we are going to add a documentation page about delphyne-template, to better explain the user where exactly to add these calls. And then we will add a reference here and in other places with a link to the delphyne-template page describing how to implement the wrapper.run() method. I might also move some content around, depending on what looks more clear. For now it's ok like this :)

Author: Do we want to prescribe how to use the `run()` method? I think the only thing that is relevant is that this is the main entry point and the one method to implement when extending the class. Then how you organise your execute calls is totally up to you.

.. code-block:: python

    self.execute_transformation(my_transformation)
    self.execute_batch_transformation(my_batch_transformation)

Raw SQL

> -------------

Reviewer: The dashed should be the same length as the title

SQL queries can easily be executed with the wrapper.
In case of just executing a simple query, the method :meth:`.Wrapper.execute_sql_query` is used.

> If the SQL query is saved in a file, the method :meth:`.Wrapper.execute_sql_file` is used.

Reviewer: Also for these, please provide a code block at the end of the paragraph showing how these calls would appear in the wrapper method

The SQL query should handle the insertion of records.

> The easiest way to create a transformations with sql is by following the template given here.

Reviewer: sql -> SQL (check the file to see if it happens anywhere else)

If you have defined your transformation in Rabbit-in-a-Hat, then
you can directly export `a SQL skeleton in this format <http://ohdsi.github.io/WhiteRabbit/RabbitInAHat.html#generating_a_sql_skeleton_(v090)>`_.
The convention is to put these transformation scripts in the `src/main/sql` folder.
`See delphyne-template for examples. <https://github.com/thehyve/delphyne-template/tree/master/src/main/sql>`_

.. code-block:: sql

    INSERT INTO @cdm_schema.<target_table> (
     <target_column1>,
     <target_column2>,
     ...
    )
    SELECT
     <source_column1>,
     <source_column2>,
     ...
    FROM @source_schema.<source_table>

SQLAlchemy query

> -------------

Reviewer: the dashes should be the same length as the title

Author: For me, it  still renders  fine though!

Author: According to these docs, it does not actually need to be the same length (or is it a local delphyne convention? 😝 )

https://docutils.sourceforge.io/docs/user/rst/quickref.html#section-structure

> Titles are underlined (**or over-
> and underlined**) with a printing
> nonalphanumeric 7-bit ASCII
> character.

Author: (still, I have updated all titles, does my OCD good...)

Reviewer: My bad, indeed it says "at least as long as".. can be longer. Sorry for being so strict 😅

Author: I just realise that the highlighted comment about over/underlining is not relevant... Can you put the characters on top of the test as well (overlined)?

Reviewer: I also didn't know it was an option 🤔 anyway so far we have used underlined, so I wouldn't change it

you can directly export `a SQL skeleton in this format <http://ohdsi.github.io/WhiteRabbit/RabbitInAHat.html#generating_a_sql_skeleton_(v090)>`_.
The convention is to put these transformation scripts in the `src/main/sql` folder.
`See delphyne-template for examples. <https://github.com/thehyve/delphyne-template/tree/master/src/main/sql>`_

.. code-block:: sql

    INSERT INTO @cdm_schema.<target_table> (
     <target_column1>,
     <target_column2>,
     ...
    )
    SELECT
     <source_column1>,
     <source_column2>,
     ...
    FROM @source_schema.<source_table>

Executing the file is then done by adding the following line to a wrapper method.
Only the filename has to be provided, delphyne will look for the file in the `src/main/sql` folder.

.. code-block:: python
        self.execute_sql_file('my_file.sql')

SQLAlchemy query
----------------
Instead of writing plain SQL query, the query can also be written using SQLAlchemy expressions.
SQLAlchemy translates the expressions into SQL.
This has the advantage that it can be compiled to any SQL dialect and makes the query agnostic of the used
Relational Database Management System (RDBMS).

> **TBC**

Reviewer: ```suggestion

.. code-block:: python

    def my_sql_transformation(wrapper):
    
        source_table = wrapper.cdm.<source_table>.__table__
        target_table = wrapper.cdm.<target_table>.__table__

        sel = select([
            source.columns['source_column_1'],
            source.columns['source_column_2'],
            ])\
            .select_from(source)

        ins = target.insert().from_select(sel.columns, sel)
        
        return ins
        
In case the source table is not part of the CDM schema, you can obtain it with the following method, which leverages SQLAlchemy's ability to create reflected table objects from the database itself:

.. code-block:: python
        source_table = wrapper.get_table(schema='my_source_schema', table_name='my_source_table')
        
Inside a wrapper method, the transformations can be called like this, similar to ORM transformations.

.. code-block:: python
    def run(self):
        ...
        self.execute_sql_transformation(my_sql_transformation)

```

    :backlinks: none

General intro
-------------

> By extending the :mod:`~src.delphyne.wrapper.Wrapper` a number of transformation functions are available.

Reviewer: you can replace this and other occurrences with :class:`.Wrapper`, it will work :)

Reviewer: btw: "by extending the Wrapper" sounds a bit weird. I would say that the Wrapper offers different ways to execute transformation scripts?

Author: Well, extending this class is what you do right?
```python
class my_wrapper(delphyne.Wrapper):
    ...
```

Author: Will still use your suggestion :)

Basically, there are two options:
 - Write a python function, returning the results as ORM objects.
 - Write a SQL query, directly inserting into the target table.

> For both options the :mod:`~src.delphyne.wrapper.Wrapper` has build-in methods that coordinate the execution.

Reviewer: builT-in (if you keep this sentence, maybe not needed)

> These methods, detailed below, use SqlAlchemy to handle database operations and log the execution statistics.

Reviewer: the capitalization is "SQLAlchemy"

Reviewer: Same for "SQL" below

Reviewer: (Check other occurrences too, it's a minor thing really but should be consistent). 
Another one I noticed: capitalization of Python (sometimes written lowercase, sometimes uppercase)

This provides a standardised interface independent of whether you are using Python or Sql for the transformation logic.

ORM
---
The ORM approach allows the use of pure python to write transformations.
This function takes the wrapper as input and produces OMOP records as SqlAlchemy ORM objects.
It can return these objects in two different ways (see below).
While records from multiple different OMOP tables can be created in the same function,

>  it is good practice to create a separate file for each table to table transformation.

Reviewer: This part is true for all transformations, move at the top before going into the details of ORM vs SQL

> See delphyne-template for examples.

Reviewer: It could be useful to link to the template repository here, specifically the `transformations` folder

Reviewer: Actually, it might be good to add specific instructions on where to place the transformations in the template:
- for ORM-based: `src/main/python/transformation`
- for SQL-based: `src/main/sql`

There are two ways of execution the transformation, each requiring a different return type.

> The first mode, used by :mod:`~src.delphyne.wrapper.Wrapper.execute_transformation`.

Reviewer: ..and this with :meth:`~.Wrapper.execute_transformation()`

Reviewer: it works irrespectively, it's just shorter if you need to edit it later in the raw .rst file

Author: Cool, that is actually very helpful to learn a bit on rst! :D

This expects the transformation function to return the full list of ORM objects at once.
All returned records are committed to the database at the same time.

While this is straightforward, it does require to store all records in memory.
For large tables there is a 'batch' mode :mod:`~src.delphyne.wrapper.Wrapper.execute_batch_transformation`.
This expects the transformation function to yield the ORM objects one at a time.
The wrapper will commits the yielded records to the database in batches of given size.

The pseudo-code for a transformation is given below, with the only difference between the two
execution modes being the way the omop records are returned.
```python

> def my_transformation(wrapper):

Reviewer: You need an empty line between the text and the start of the code block, otherwise it doesn't display correctly. Also between `python` and the first line of code.

    # read source data
    for row in source:
        # transformation logic...
        omop_record = OmopTable(  # e.g. Person, Observation, ...
            person_id=...,
            concept_id=...,
            ...
        )
        # yield OR store in list
    # return list (if not yielding)

> ```

Reviewer: I suggest providing separate explicit examples for the case with return and yield, and add blocks showing how you would execute the corresponding transformation with either `execute_transformation` or `execute_batch_transformation` in the Wrapper. I think you can use of the the OMOP tables explicitly, say `Person`

Raw SQL
-------------

> SQL queries can easily be executed with through the wrapper.

Reviewer: "with" or "through"

In case of just executing a simple query, the method :mod:`~src.delphyne.wrapper.Wrapper.execute_sql_query` is used.

> If the sql query is saved in a file, the method :mod:`~src.delphyne.wrapper.Wrapper.execute_sql_file` is used.

Reviewer: Also in this case, could be good to write out an example. There is a difference with `execute_transformation()`, that is the full SQL file name has to written out explicitly as a string (e.g. `execute_sql_file('my_file.sql'`))

The sql query should handle the insertion of records.
The easiest way to create a transformations with sql is by following the template given here.
If you have defined your transformation in Rabbit-in-a-Hat, then
 you can directly export `a sql skeleton in this format <http://ohdsi.github.io/WhiteRabbit/RabbitInAHat.html#generating_a_sql_skeleton_(v090)>`_.

```sql

> INSERT INTO @cdmSchma.<target_table> (

Reviewer: typo in `@cdmSchema`

Reviewer: actually, the placeholder name should be `@cdm_schema`. this will be taken straight from the `schema_translate_map` section in the configuration file

 <target_column1>,
 <target_column2>,
 ...
)
SELECT
 <source_column1>,
 <source_column2>,
 ...

> FROM @sourceSchema.<source_table>

Reviewer: I was wondering if we could give an example where the data is read from file rather than from another schema, since that's our main use case.

Author: For pure SQL, that is not our main use case, right?

> ```

Reviewer: same here, you will need something like `.. code-block:: sql` (see if it works, I haven't tried that one)
